src/GeneratedMapper/MapperGenerator.cs

```diff
@@ -56,6 +59,32 @@ public void Execute(GeneratorExecutionContext context)
 
                     context.AddSource(name, text);
                 }
+
+                var lines = string.Join(Environment.NewLine, foundMappings.Where(x => x.SourceType != null && !x.AfterMaps.Any()).Select(x =>
+                {
+                    return
+                        $@"                case {x.SourceType.ToDisplayString()} val:
+                    return {x.SourceType.ContainingNamespace.ToDisplayString()}.{x.SourceType.Name}MapToExtensions.MapTo{x.DestinationType.Name}(val) is TDestination destination ? destination : default;";
+                }));
+                var text2 =
+                    $@"using System;
+
+namespace GeneratedMapper.Extensions
+{{
+    public static class MapExtensions
+    {{
+        public static TDestination MapTo<TSource, TDestination>(this TSource source)
+        {{
+            switch (source)
```

Comment:
I don't really like this switch here. If we have like 100 mappings, this switch is going to be huge. And this can lead to runtime issues when you put in a type that isn't supported, you'll get a `NotSupportedException` and no compile error. Could you generate like `MapTo<TSource, TDestination>` methods that have a `where` configure so you pin what types you can pick and the developer sees during development / build that some mapping is having a problem.

Reply:
I tried that but when I did Resharper gave ambiguous method call couldn't determine which one it was.  Maybe once the new one is created you then have to disable the main one somehow from being used for the code generated, which might be possible by I don't know how you would dynamically filter out generic conditions, so this isn't the case.
- If you know how to make it not ambiguous I'm all ears, but for now the only way I got it to work is making this class part of the code generation process.
- Need to start with something people can call, and when they call add on top of it so their case works.

The idea is that if you call `MapTo<TSource,TDestination>` at all it will make a case for said types no matter what, so you'll never hit the runtime not supported exception.  The issue comes into play adding Maps that have AfterMap or Extensions with constructor parameters and how we inform the user.

1. Put the code into the switch statement and let the compilation fail with Argument count miss-match happen in the MappingExpression.g.cs
2. Identify the miss match in validation and then add Error to diagnostics explaining the resolvers/extensions are blocking them from using code in `MapTo<TSource,TDestination>` function, and have the source be the call in question vs the auto generated code.
  - Also question of putting the code in there as well, so you will get both exceptions, but the first will be the custom one we make which they should solve first.  That's usually how compilers work.  Fix first thing and fixes a bunch of other smaller issues that were inferred incorrectly from the original problem.

Reply:
Yeah after giving it more thought I realized that this is a C# syntax limitation. You can't make overloads where only the `where`s are different. So this is something we have to accept.

For the issue with the parameters, that is going to be quite a challenge. You have one single `MapTo<,>` name you want to use for potentially all the mapping methods, which could have a large variety of parameter combinations. Solving that is perhaps a bit too much for the initial implementation. Perhaps we should view this `MapTo<,>` syntax for now an ad-hoc variant, where there is no support for parameters yet.